### PR TITLE
Media: Add a settings dialog to the Media Library toolbar.

### PR DESCRIPTION
**Pull request by t-hamano**
- Alternatives to #12795
- Trac ticket: https://core.trac.wordpress.org/ticket/65775

## Overview

WordPress core does not yet seem to have an API for rendering what are known as popovers. This PR instead attempts to render a settings dialog for the media library using the HTML native `dialog` element. The only concern is that this PR is the first to introduce a UI using the dialog element.

## New APIs

- PHP:
  - `wp_ajax_set_media_library_settings()`
- JavaScript
  - `wp.media.view.LibrarySettings`
  - `wp.media.view.settings.librarySettings`

## Use of AI Tools

> [!NOTE]
> This PR is a prototype. Almost all of the code in this PR was generated by Claude. If we find the approach taken in this PR to be reasonable, I would like to review all of the code.

## Screenshots

image
image
image

### CODE REVIEW

**Comment by joedolson** [comment]
I pushed a few minor visual tweaks and changed the `label` configuration to avoid using an implicit label. Overall, I think this is very workable.

I noticed that the modal is not actually restricting focus, so that will need to be adjusted, but this is going in a direction that I think is viable.

File: src/wp-includes/media.php [hunk 1]
```diff
@@ -5104,6 +5101,12 @@ function wp_enqueue_media( $args = array() ) {
 		'months'            => $months,
 		'mediaTrash'        => MEDIA_TRASH ? 1 : 0,
 		'infiniteScrolling' => ( $infinite_scrolling ) ? 1 : 0,
+		// Data for the settings dialog. Unlike the value above, `infiniteScrolling` here is the unfiltered personal option the dialog edits.
+		'librarySettings'   => array(
+			'nonce'             => wp_create_nonce( 'media-library-settings' ),
+			'infiniteScrolling' => ( $user_infinite_scrolling ) ? 1 : 0,
+			'isFiltered'        => ( $infinite_scrolling !== $user_infinite_scrolling ) ? 1 : 0,
+		),
```
[the remark this answers is not in the document]
Comment (t-hamano) [reply src/wp-includes/media.php]: fixed in 2b58600768a78ac5d23827c96fa9c88bec41ad16

File: src/js/media/views/library-settings.js [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+	 * @return {void}
+	 */
+	open: function() {
+		if ( ! this.dialog ) {
+			this.createDialog();
+		}
+
+		this.setStatus( '' );
+		this.dialog.showModal();
+	},
```
[the remark this answers is not in the document]
Comment (t-hamano) [reply src/js/media/views/library-settings.js:110]: This fallback is likely unnecessary, as `showModal` has been supported by all major browsers for several years.

File: src/wp-includes/media-template.php [hunk 3]
```diff
@@ -344,6 +344,50 @@ function wp_print_media_templates() {
...
+				<p class="media-library-settings__field">
+					<input type="checkbox" class="media-library-settings__checkbox"
+						id="{{ data.infiniteScrollingId }}"
+						aria-describedby="{{ data.infiniteScrollingId }}-description" autofocus
+						<# if ( data.infiniteScrolling ) { #>checked<# } #> />
+					<label for="{{ data.infiniteScrollingId }}">
+						<?php _e( 'Infinite scrolling' ); ?>
+					</label>
+				</p>
+				<p class="description" id="{{ data.infiniteScrollingId }}-description"><?php _e( 'Load more media items automatically when scrolling, instead of using a Load more button.' ); ?></p>
```
[the remark this answers is not in the document]
Comment (t-hamano) [reply src/wp-includes/media-template.php:385]: Fixed in d280e56de0e83a82aec44c5b1e1952b053bf434d

File: src/wp-admin/includes/ajax-actions.php [hunk 4]
```diff
@@ -1891,6 +1891,30 @@ function wp_ajax_update_welcome_panel() {
...
+		wp_send_json_error( array( 'message' => __( 'The setting could not be saved.' ) ), 400 );
+	}
+
+	$infinite_scrolling = 'true' === wp_unslash( $_POST['infinite_scrolling'] ) ? 'true' : 'false';
+
+	// Plain user meta, without the site prefix update_user_option() would add, to match the personal option on the profile screen.
+	update_user_meta( get_current_user_id(), 'infinite_scrolling', $infinite_scrolling );
+
+	wp_send_json_success( array( 'infiniteScrolling' => 'true' === $infinite_scrolling ) );
+}
```
[the remark this answers is not in the document]
Comment (t-hamano) [reply src/wp-admin/includes/ajax-actions.php:1915]: Fixed in a10841e0154a65e2b63213b3a3bb4512d13b81b8

File: src/wp-admin/css/media.css [hunk 5]
```diff
@@ -571,7 +571,14 @@ border color while dragging a file over the uploader drop area */
 	min-height: 80px;
 }
 
-.media-frame.mode-grid .media-toolbar label:not(.media-search-input-label) {
+/* Allows space for the display settings toggle in narrow viewports. */
+#media-search-input {
+	max-width: calc( 100% - 40px );
+}
```
[the remark this answers is not in the document]
Comment (t-hamano) [reply src/wp-admin/css/media.css]: Fixed in 837d3b4442963e6ec2a09c36a4804bba2e9afae5

**Comment by t-hamano** [comment]
I believe I've addressed all the issues I could think of. We should be ready for review now.

**Comment by joedolson** [comment]
Pushed two changes on narrower viewports - the button position below 782px and the button size between 782px and 900px.

File: src/wp-includes/media-template.php [hunk 6]
```diff
@@ -344,6 +344,50 @@ function wp_print_media_templates() {
...
+					<?php // The checked state is set by the view each time the dialog is opened. ?>
+					<input type="checkbox" class="media-library-settings__checkbox"
+						id="{{ data.infiniteScrollingId }}"
+						aria-describedby="{{ data.infiniteScrollingId }}-description" autofocus />
+					<label for="{{ data.infiniteScrollingId }}">
+						<?php _e( 'Enable infinite scrolling' ); ?>
+					</label>
+				</p>
+				<p class="description" id="{{ data.infiniteScrollingId }}-description"><?php _e( 'Load more media items automatically when scrolling, instead of using a Load more button.' ); ?></p>
+				<p class="media-library-settings__status" role="status"></p>
```
Comment (joedolson) [inline src/wp-includes/media-template.php:386]: The role of 'status' is equivalent to `aria-live="polite"`, and I don't think that's quite right here. Mostly because of the extended `aria-description`, which means that when you toggle the control, you have to wait for the full label & description to finish reading before the status announcement occurs. For that reason, I'd be more inclined to use `role="alert"`.

@afercia Do you have an opinion here?

File: src/wp-includes/css/media-views.css [hunk 7]
```diff
@@ -984,6 +984,102 @@ select#media-attachment-filters ~ select#media-attachment-date-filters {
 	display: block;
 }
 
+/**
+ * Media Library settings
+ */
+.wp-core-ui .button.media-library-settings__toggle {
```
Comment (joedolson) [inline src/wp-includes/css/media-views.css:990]: Having the button visually embedded in the search field is something I'm not really sure about. It somewhat conveys the idea that this is specifically *search* options, rather than general media library options.

**Review by joedolson — APPROVED** [review]
I have two items that I think are non-blocking changes, and can be iterated based on feedback.

**Comment by annezazu** [comment]
I don’t think this is very discoverable or well connected to the media library. As Joe notes above, it feels more about search options. I’m not comfortable right now proceeding with this, especially with where we are in the release cycle too.

**Comment by joedolson** [comment]
@annezazu Honestly, the positioning of the button is a relatively trivial part of this; relocating that is very doable. The more important part, in my opinion, is:

- Is a settings toggle a good idea? (In my opinion yes, this is an appropriate place for this.)
- Is the dialog functional, accessible, and clear? (Also in my opinion, yes.)

But I'm not convinced by the location; I think it was better when it wasn't directly inlined with the search field, however, and think it probably should be even earlier in the modal.

**Comment by t-hamano** [comment]
> it feels more about search options

We could indicate that the button is an independent option by placing everything except the toggle button on the left side, but I believe this would require a significant CSS change. I believe the structure of the media toolbar needs a fundamental overhaul.

image

**Comment by fcoveram** [comment]
The button location shown in @t-hamano's message above looks good to me. The layout is similar to that in DataViews (Picker) and the media upload UX.

**Comment by tyxla** [comment]
@jasmussen I just requested review from you here but saw @fcoveram's feedback late. Thank you both!

**Comment by jasmussen** [comment]
Happy to ship this as well! Thanks all.

**Comment by t-hamano** [comment]
> We could indicate that the button is an independent option by placing everything except the toggle button on the left side, but I believe this would require a significant CSS change. I believe the structure of the media toolbar needs a fundamental overhaul.

To be honest, I'm hesitant to make major CSS changes around the media view and media toolbar right now because the CSS is overly complex 😅 While we might refactor the media-related UI with modern CSS in the future, I don't want to do that right before the final RC release. Let's find an acceptable compromise for the 7.1 release.

File: src/js/media/views/attachments.js [hunk 8]
```diff
@@ -114,6 +114,28 @@ Attachments = View.extend(/** @lends wp.media.view.Attachments.prototype */{
...
+	setInfiniteScrolling: function( enabled ) {
+		this.options.infiniteScrolling = enabled;
+
+		$( this.options.scrollElement ).off( 'scroll', this.scroll );
+
+		if ( enabled ) {
+			$( this.options.scrollElement ).on( 'scroll', this.scroll );
+
+			// The list may already be scrolled past the point where more are loaded.
+			this.scroll();
```
Comment (tyxla) [inline src/js/media/views/attachments.js:135]: Noting that `this.scroll()` never checks `this.options.infiniteScrolling` so if there's already a queued throttled callback or pending `collection.more()`, they might call `scroll()` again, which will run unchecked. We might want to add a `this.options.infiniteScrolling` check to `scroll()` itself.

File: src/js/media/views/library-settings.js [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+
+	/**
+	 * Whether a `media_library_infinite_scrolling` filter callback overrides the
+	 * personal option.
+	 *
+	 * @type {boolean}
+	 */
+	isFiltered: !! settings.librarySettings &&
+		Boolean( settings.infiniteScrolling ) !== Boolean( settings.librarySettings.infiniteScrolling ),
```
Comment (tyxla) [inline src/js/media/views/library-settings.js]: Hmm, aren't we risking losing the filter precedence here? Disabling the checkbox here changes the UI to "Load More" despite the filter requiring infinite scrolling.

File: src/js/media/views/library-settings.js [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+	 */
+	updateInfiniteScrolling: function( event ) {
+		var view = this,
+			checkbox = event.target,
+			enabled = checkbox.checked,
+			requestId = ++this.requestId;
+
+		this.setStatus( __( 'Saving…' ) );
+
+		wp.ajax.post( 'set-media-library-settings', {
```
Comment (tyxla) [inline src/js/media/views/library-settings.js:145]: Multiple concurrent saves can result in saving the wrong value. Might need some throttling or disabling if a request is in progress here.

File: src/js/media/views/library-settings.js [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+			if ( ! view.isFiltered ) {
+				settings.infiniteScrolling = enabled ? 1 : 0;
+
+				view.controller.trigger( 'library:infinite-scrolling', enabled );
+			}
+
+			view.setStatus( enabled ?
+				__( 'Infinite scrolling is on.' ) :
+				__( 'Infinite scrolling is off.' )
+			);
```
Comment (tyxla) [inline src/js/media/views/library-settings.js:172]: Is this yet another place that might be out of sync with `isFiltered`? Looks like we're not covering the case when a filter has forced that infinite scroll is "on".

File: src/js/media/views/attachments/browser.js [hunk 12]
```diff
@@ -111,6 +115,39 @@ AttachmentsBrowser = View.extend(/** @lends wp.media.view.AttachmentsBrowser.pro
...
+	 *
+	 * @return {void}
+	 */
+	setInfiniteScrolling: function( enabled ) {
+		if ( enabled === Boolean( this.attachments.options.infiniteScrolling ) ) {
+			return;
+		}
+
+		this.attachments.setInfiniteScrolling( enabled );
+		this.$el.toggleClass( 'has-load-more', ! enabled );
```
Comment (tyxla) [inline src/js/media/views/attachments/browser.js:136]: We might need to cover a few more items/classes based on this logic, like `.more-loaded`, `.found-media`, and `.new-media` for example.

File: src/js/media/views/library-settings.js [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+	__ = wp.i18n.__,
+	LibrarySettings;
+
+/**
+ * wp.media.view.LibrarySettings
+ *
+ * A toolbar control opening a modal dialog with the personal options for the
+ * Media Library.
+ *
+ * @since 7.1.0
```
Comment (tyxla) [inline src/js/media/views/library-settings.js:13]: Considering the 7.1 branch is only for bug fixes at this point, I doubt if this will make it. Should we consider it for 7.2?

File: src/wp-includes/css/media-views.css [hunk 14]
```diff
@@ -984,6 +984,102 @@ select#media-attachment-filters ~ select#media-attachment-date-filters {
...
+	position: absolute;
+	right: 0;
+}
+
+.wp-core-ui .button.media-library-settings__toggle .dashicons {
+	line-height: 1;
+}
+
+.wp-core-ui .media-toolbar-primary > .media-library-settings__toggle {
+	margin-left: 8px;
```
Comment (tyxla) [inline src/wp-includes/css/media-views.css]: Isn't the toggle absolute? Why use margins?

File: src/wp-includes/css/media-views.css [hunk 15]
```diff
@@ -984,6 +984,102 @@ select#media-attachment-filters ~ select#media-attachment-date-filters {
...
+}
+
+.wp-core-ui .mode-grid .media-toolbar-primary > .media-library-settings__toggle {
+	margin-left: 0;
```
Comment (tyxla) [inline src/wp-includes/css/media-views.css]: Same question as above.

File: src/js/media/views/attachments/browser.js (hunk 12 above)
Comment (tyxla) [inline src/js/media/views/attachments/browser.js:136]: Wouldn't enabling infinite scrolling in the modal always flash the spinner and fetch an extra page? Should we swap those 2 lines to address that?

**Review by tyxla — COMMENTED** [review]
Thanks for working on this @t-hamano 🙌 The feature is a nice quality-of-life win.

My main hesitation is timing rather than the code. We're at RC2, and this adds a new public JS API, a new AJAX endpoint, and runtime layout changes to one of the most-exercised screens in the admin, one that a lot of plugins hook into. The issues I highlight in review are in runtime the happy path, which makes me suspect there's more we haven't found yet.

Could we get some broad manual testing before this lands? Toggling the setting mid-session in both grid mode and the modal, across browsers, with filters and search active, and with at least one plugin that customises the media views. If that comes back clean, great. But if it feels like it needs more than a quick pass it could very well be a 7.2 item.

File: src/js/media/views/attachments.js (hunk 8 above)
Comment (t-hamano) [reply src/js/media/views/attachments.js:135]: Fixed in 431a1b9f669f8d169521ef4eebff11a006795e7a

File: src/js/media/views/library-settings.js (hunk 9 above)
Comment (t-hamano) [reply src/js/media/views/library-settings.js]: Fixed in cf528d3536b2e1c23ad0484c3c5602923f84aa50. I made it explicit whether a filter exists using `has_filter`.

File: src/js/media/views/library-settings.js (hunk 11 above)
Comment (t-hamano) [reply src/js/media/views/library-settings.js:172]: I am considering this, and I have two proposals.

- Notify that the filter is enabled when settings are updated. For example, `Setting saved. Infinite scrolling is currently controlled by a plugin or theme.`
- When filtered, the toggle button should not be displayed at all.

File: src/js/media/views/library-settings.js (hunk 10 above)
Comment (t-hamano) [reply src/js/media/views/library-settings.js:145]: Fixed in 93997bab9aa87c18bfd9e4ffae66afc134898dbd

File: src/js/media/views/attachments/browser.js (hunk 12 above)
Comment (t-hamano) [reply src/js/media/views/attachments/browser.js:136]: Fixed in 8562d2aaa552f20f3c054f8bd3d8dde4a42bed69

File: src/js/media/views/library-settings.js (hunk 13 above)
Comment (t-hamano) [reply src/js/media/views/library-settings.js:13]: This will need to be agreed upon and decided among the release leads.

File: src/wp-includes/css/media-views.css (hunk 14 above)
Comment (t-hamano) [reply src/wp-includes/css/media-views.css]: Nice catch, this is something I forgot to remove when I absolutely positioned the toggle. Removed in cd6074aa525bc487814afce65ca3ae3a7bab7943

File: src/wp-includes/css/media-views.css (hunk 15 above)
Comment (t-hamano) [reply src/wp-includes/css/media-views.css]: Removed in https://github.com/WordPress/wordpress-develop/commit/cd6074aa525bc487814afce65ca3ae3a7bab7943

**Comment by t-hamano** [comment]
@tyxla Thanks for the review!

> My main hesitation is timing rather than the code. We're at RC2, and this adds a new public JS API, a new AJAX endpoint, and runtime layout changes to one of the most-exercised screens in the admin, one that a lot of plugins hook into. The issues I highlight in review are in runtime the happy path, which makes me suspect there's more we haven't found yet.

I agree with this. While I understand the importance of improving accessibility, there is also a risk of causing problems by introducing new APIs.

This might go against [the decision](https://github.com/WordPress/wordpress-develop/pull/12795#issuecomment-5155341092), but personally I propose punting this PR to 7.2. There is too little time to prepare.

cc @annezazu @joedolson

**Comment by t-hamano** [comment]
> Could we get some broad manual testing before this lands? Toggling the setting mid-session in both grid mode and the modal, across browsers, with filters and search active, and with at least one plugin that customises the media views.

I tested the layout changes when the Media Library Assistant plugin, which extends the media toolbar, was enabled.

<details><summary>Before</summary>

before-desktop
before-tablet
before-mobile
image

</details> 

<details><summary>After</summary>

after-desktop
after-tablet
after-mobile
image

</details>

File: src/js/media/views/library-settings.js (hunk 11 above)
Comment (joedolson) [reply src/js/media/views/library-settings.js:172]: To me, the first option is more clear.

**Comment by 369work** [comment]
Test in Playground

I tested PR #12932 for the first time, and I don't think it would be immediately obvious to users that they need to click the gear icon.

I think the “Enable infinite scrolling” checkbox could be placed somewhere more visible and intuitive, so users can discover the setting more easily.


スクリーンショット 2026-08-12 003316
スクリーンショット 2026-08-12 004215

**Comment by annezazu** [comment]
Coming here to say that I agree we need to punt this from the 7.1 release. I am just uncomfortable with these kinds of changes at this stage of the release process when we need utmost stability. I also think this is worthy of being considered for a point release for 7.1 and that's something we can pursue without waiting months until 7.2.
